patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
++import io.papermc.paper.command.brigadier.argument.ArgumentResolver;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.command.CommandSender;
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Represents a vanilla command source stack which is used
++ * for command execution.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface CommandSourceStack {
++
++    /**
++     * Gets the bukkit entity running this command.
++     * @return bukkit entity, null if not present
++     */
++    @Nullable
++    Entity getBukkitEntity();
++
++    /**
++     * Gets the world that this command is being executed in.
++     * @return world
++     */
++    @Nullable
++    World getBukkitWorld();
++
++    /**
++     * Gets the location that this command is being executed in.
++     * @return location
++     */
++    @Nullable
++    Location getBukkitLocation();
++
++    /**
++     * Gets the sender representing this source stack.
++     * @return source stack sender
++     */
++    @NotNull
++    CommandSender getBukkitSender();
++
++    /**
++     * Resolves an argument in this command source stack.
++     *
++     * @param context command context
++     * @param name name of argument
++     * @return resolved argument
++     * @param <V> resolved argument type
++     */
++    @NotNull
++    @SuppressWarnings("unchecked")
++    default <V> V getResolvedArgument(@NotNull CommandContext<CommandSourceStack> context, @NotNull final String name) throws CommandSyntaxException {
++        ArgumentResolver<?> resolver = context.getArgument(name, ArgumentResolver.class);
++        return (V) resolver.resolve(this);
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..94305f3f35d5f807381dcd3b26200f86cb0816e8
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
+@@ -0,0 +1,33 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.Message;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.serializer.ComponentSerializer;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.ServiceLoader;
++
++/**
++ * A component serializer that converts it to a brigadier compatible type.
++ */
++public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
++
++    /**
++     * A component serializer that converts it to a brigadier compatible type.
++     *
++     * @return serializer instance
++     */
++    @NotNull
++    static MessageComponentSerializer message() {
```

Comment:
If we exposed this via PaperComponents, we could avoid an extra `class Holder`. It is component related for sure.

Reply:
This would be different from every other serializer, which has a Holder class. I do not really see the issue with the extra class.